lib/internal/url.js

```diff
@@ -0,0 +1,328 @@
+  }
+
+  get protocol() {
+    return this[context].scheme;
+  }
+
+  set protocol(scheme) {
+    if (typeof scheme !== 'string')
+      throw new TypeError('protocol must be a string');
+    binding.parse(scheme, binding.kScheme, null, (...args) => {
+      // TODO: handle failure cases
+      this[context].scheme = args[binding.ARG_PROTOCOL];
+    });
+  }
+
+  get username() {
+    return this[context].username;
+  }
+
+  set username(username) {
+    if (typeof username !== 'string')
+      throw new TypeError('username must be a string');
+    this[context].username = encodeAuth(username);
+  }
+
+  get password() {
+    return this[context].password;
+  }
+
+  set password(password) {
+    if (typeof password !== 'string')
+      throw new TypeError('password must be a string');
+    this[context].password = encodeAuth(password);
+  }
+
+  get host() {
+    var ret = this[context].host;
+    if (this[context].port !== undefined)
+      ret += `:${this[context].port}`;
+    return ret;
+  }
+
+  set host(host) {
+    if (typeof host !== 'string')
+      throw new TypeError('host must be a string');
+    if (this[cannotBeBase])
+      return;
+    binding.parse(host, binding.kHost, null, (...args) => {
+      // TODO: handle failure cases
+      this[context].host = args[binding.ARG_HOST];
+      if (args[binding.ARG_PORT] !== undefined)
+        this[context].port = args[binding.ARG_PORT];
+    });
+  }
+
+  get hostname() {
+    return this[context].host;
+  }
+
+  set hostname(host) {
+    if (typeof host !== 'string')
+      throw new TypeError('host must be a string');
+    if (this[cannotBeBase])
+      return;
+    binding.parse(host, binding.kHostname, null, (...args) => {
+      // TODO: handle failure cases
+      this[context].host = args[binding.ARG_HOST];
+    });
+  }
+
+  get port() {
+    return this[context].port;
+  }
+
+  set port(port) {
+    if (!this[context].host || this[cannotBeBase] || this.protocol === 'file:')
+      return;
+    port |= 0;
+    if (port < 0 || port > 0xffff)
+      throw new RangeError('port must be a number between 0 and 65535');
+    this[context].port = port;
+  }
+
+  get pathname() {
+    if (this[cannotBeBase])
+      return this[context].path[0];
+    return this[context].path !== undefined ?
+      `/${this[context].path.join('/')}` : '';
+  }
+
+  set pathname(path) {
+    if (this[cannotBeBase])
+      return;
+    binding.parse(path, binding.kPathStart, null, (...args) => {
+      // TODO: handle failure cases
+      this[context].path = args[binding.ARG_PATH];
+    });
+  }
+
+  get search() {
+    return this[context].query === undefined ? '' : `?${this[context].query}`;
+  }
+
+  set search(search) {
+    if (!search) {
+      // TODO: clear the search params object also
+      delete this[context].query;
+      return;
+    }
+    if (typeof search !== 'string')
+      throw new TypeError('search must be a string');
```

Comment:
When you change this to a coercion, be sure to move it _before_ the `!search` check. `.search = 0` should not bail out early; it should be equivalent to `.search = "0"`.